subsys/net/lib/sockets/sockets.c

```diff
@@ -1497,6 +1497,19 @@ int zsock_getsockopt_ctx(struct net_context *ctx, int level, int optname,
 	switch (level) {
 	case SOL_SOCKET:
 		switch (optname) {
+		case SO_TYPE: {
+			int type = (int)net_context_get_type(ctx);
+
+			if (*optlen >= sizeof(int)) {
+				*(int *)optval = type;
+				*optlen = sizeof(int);
+			} else {
+				memcpy(optval, &type, *optlen);
+			}
```

Comment:
The expected value of the type is int, so wouldn't something like this work fine here?
```
if (*optlen != sizeof(int)) {
    errno = EINVAL;
    return -1;
}

*(int *)optval = type;
return 0;
```
Would there be a usecase where user could supply non-int pointer and/or non-int optlen?

Reply:
It would work perfectly fine and that's how I would define this API. However POSIX defines it like this:

> If the size of the option value is greater than option_len, the value stored in the object pointed to by the option_value argument shall be silently truncated. Otherwise, the object pointed to by the option_len argument shall be modified to indicate the actual length of the value.

I understand it's preferred to follow POSIX, instead of improving it. Am I correct?

Reply:
On the other hand, Linux defines it like you suggest:

> EINVAL optlen invalid in setsockopt().  In some cases this error can also occur for an invalid value in optval (e.g., for the IP_ADD_MEMBERSHIP option described in ip(7)).

Shall we follow POSIX or Linux? What's the rule of thumb here?

Reply:
I would follow what Linux does here. The returned value should be usable after a successfull call, and if the value is truncated, it is not. If the optlen is longer than the size, it also is quite confusing as how the caller of `getsockopt()` we be able to interpret the value then. @pfalcon do you have opinion about this?

Reply:
I'd say this is a rather corner case. So, my suggestions would be:

1. To follow what our existing code does in the similar cases (getsockopt specifically).
2. And I bet our existing code is optimized for size, and so `if (*optlen != sizeof(int)) EINVAL` makes good sense.

If we ever get a report of issues with real-world app, we can revisit. (But I guess we unlikely would, and the only situation we might hit this is if running some "strict POSIX compliance testsuite". But it would discover a whole bunch of other issues before that, and fixing them all would be a separate project).

Reply:
I'll update the code in the evening. I like the `EINVAL` solution much more than POSIX one!